resources/testharness.js

```diff
@@ -587,11 +587,12 @@ policies and contribution forms [3].
 
     function async_test(func, name, properties)
     {
-        if (tests.promise_setup_called) {
-            tests.status.status = tests.status.ERROR;
-            tests.status.message = '`async_test` invoked after `promise_setup`';
-            tests.complete();
+        // async_test are deferred when a promise_setup is pending.
+        if (tests.promise_setup_pending >= 1) {
+            tests.async_test_pending.push(arguments);
+            return;
         }
```

Comment:
I'm concerned about the fact that this delays registering the test until later, versus `promise_test` which registers the test immediately but delays running it until its spot in `tests.promise_tests`. I can't however quite articulate what might go wrong with this approach; I fear non-deterministic test ordering but am not sure I can create a reproduction.

The more consistent method, to me, would be to create the `Test` object and then defer everything inside `if (func)` until later.

Reply:
I believe this is a good suggestion. Thanks!
Please let me some time to address this.

Moreover, I just realized you have unittest about testharness.js. So I will have to add a few and check the existing ones.

Reply:
Done. Please take another look.
I also added ~10 tests.